test/e2e/logging_test.go

```diff
@@ -0,0 +1,183 @@
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/pkg/system"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/logstream"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	"knative.dev/serving/pkg/network"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const request_log_template = `
+{"requestMethod": "{{.Request.Method}}", "requestUrl": "{{js .Request.RequestURI}}", "userAgent": "{{js .Request.UserAgent}}"}
+`
+
+func TestRequestLogs(t *testing.T) {
+	t.Parallel()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+
+	// Patch config-observability to enable request logs
+	obcm, err := rawCM(clients, "config-observability")
+	if err != nil {
+		t.Errorf("Error retrieving observability configmap: %v", err)
+	}
+
+	patchedObcm := obcm.DeepCopy()
+	patchedObcm.Data["logging.request-log-template"] = request_log_template
+	patchedObcm.Data["logging.enable-probe-request-log"] = "true"
+	patchCM(clients, patchedObcm)
+	defer patchCM(clients, obcm)
+
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	t.Log("Creating a new Service")
+
+	resources, err := v1test.CreateServiceReady(t, clients, &names, []rtesting.ServiceOption{
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey: "1",
+			autoscaling.MaxScaleAnnotationKey: "1",
+		})}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	url := resources.Route.Status.URL.URL()
+	if _, err := pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		url,
+		v1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(test.HelloWorldText))),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https),
+	); err != nil {
+		t.Fatalf("The endpoint %s for Route %s didn't serve the expected text %q: %v", url, names.Route, test.HelloWorldText, err)
+	}
+
+	pod, err := theOnlyPod(clients, resources.Revision.Namespace, resources.Revision.Name)
+	if err != nil {
+		t.Fatalf("Fail to fetch the pod: %v", err)
+	}
+
+	// A request was sent to / in WaitForEndpointState.
+	if err := waitForLog(clients, pod.Namespace, pod.Name, "queue-proxy", func(log map[string]interface{}) bool {
+		if v, ok := log["requestUrl"]; !ok {
+			return false
+		} else if v.(string) != "/" {
+			return false
+		}
+
+		if v, ok := log["userAgent"]; !ok {
+			return false
+		} else {
+			return v.(string) != network.QueueProxyUserAgent
+		}
```

Comment:
no need for else here.

Reply:
`v` is in the scope of `if`? If no else, we need to split if and its condition?

Reply:
ish, linter suggests initializing on its own line.